File: 2015/day2.js
const fs = require('fs')
const myInput = fs.readFileSync('./input/day2.txt', 'utf8').trim()
const myBoxes = myInput.split("\n")

const useDimensions = (input, callback) => input.split("\n").reduce((total, box) => {
    const dimensions = box.split('x').map(Number).sort((a, b) => a - b)
    return total + callback(...dimensions)
}, 0)
const wrap = (a, b, c) => 2 * (b * c + a * c) + 3 * a * b

console.log(useDimensions(myInput, wrap))

const ribbon = (a, b, c) => a * b * c + 2 * a + 2 * b
console.log(useDimensions(myInput, ribbon))

File: 2015/day2.rb
input = File.read('./input/day2.txt')

wrapping = input.split("\n").reduce(0) do |total, line|
  sides = line.split('x').map(&:to_i)
  areas = sides.combination(2).map{ |a, b| a * b }
  extra = areas.min
  total + areas.reduce(extra) { |total, area| total + 2 * area}
end

puts wrapping

ribbon = input.split("\n").reduce(0) do |total, line|
  sides = line.split('x').map(&:to_i)
  ribbon = sides.combination(2).map{ |a, b| a + b }.min * 2
  bow = sides.reduce(1) {|total, side| total * side}
  total + ribbon + bow
end

puts ribbon
